source/common/upstream/resource_manager_impl.h

```diff
@@ -44,24 +47,41 @@ class ResourceManagerImpl : public ResourceManager {
 private:
   struct ResourceImpl : public Resource {
     ResourceImpl(uint64_t max, Runtime::Loader& runtime, const std::string& runtime_key,
-                 Stats::BoolIndicator& circuit_breaker_open)
+                 Stats::BoolIndicator& circuit_breaker_open, Stats::Gauge& remaining)
         : max_(max), runtime_(runtime), runtime_key_(runtime_key),
-          circuit_breaker_open_(circuit_breaker_open) {}
+          circuit_breaker_open_(circuit_breaker_open), remaining_(remaining) {
+      remaining_.set(max);
+    }
     ~ResourceImpl() { ASSERT(current_ == 0); }
 
     // Upstream::Resource
     bool canCreate() override { return current_ < max(); }
     void inc() override {
       current_++;
       circuit_breaker_open_.set(!canCreate());
+      updateRemaining();
     }
     void dec() override {
       ASSERT(current_ > 0);
       current_--;
       circuit_breaker_open_.set(!canCreate());
+      updateRemaining();
     }
     uint64_t max() override { return runtime_.snapshot().getInteger(runtime_key_, max_); }
 
+    /**
+     * We set the gauge instead of incrementing and decrementing because,
+     * though atomics are used, it is possible for the current resource count
+     * to be greater than the supplied max.
+     */
+    void updateRemaining() {
+      /**
+       * We cannot use std::max here because max() and current_ are
+       * unsigned and subtracting them may overflow.
+       */
+      remaining_.set(max() > current_ ? max() - current_ : 0);
```

Comment:
Hmm, I think you actually would want to do something like
```
const uint64_t current_copy = current_;
remaining_set.set(max_ > current_copy ? max_ - current_copy : 0);
```

otherwise it's possible for `current_` to change between the time you check it vs the time you set it.